### PR TITLE
feat: [CI-9787]: internet connectivity check in healthcheck

### PR DESCRIPTION
**Pull request by devkimittal**

### CODE REVIEW

File: handler/health.go [hunk 1]
```diff
@@ -20,6 +22,30 @@ func HandleHealth() http.HandlerFunc {
...
 		}
-		WriteJSON(w, response, http.StatusOK)
+		status := http.StatusOK
+
+		if !checkInternetConnectivity() {
+			response = api.HealthResponse{
+				Version: version,
+				OK:      false,
+			}
+			status = http.StatusInternalServerError
```
Comment (vistaarjuneja) [inline handler/health.go:32]: can we return a better error code here? And also, there should be some error message that we can log on the client side.

File: handler/health.go [hunk 2]
```diff
@@ -20,6 +22,30 @@ func HandleHealth() http.HandlerFunc {
...
+		}
+
+		WriteJSON(w, response, status)
+	}
+}
+
+func checkInternetConnectivity() bool {
+	dialer := net.Dialer{
```
Comment (vistaarjuneja) [inline handler/health.go:40]: I am wary about this change.
What if VMs are coming up but the network is slow. In that case, connectivity check will always fail here and previously passing health checks will now start to fail.

File: handler/health.go [hunk 3]
```diff
@@ -20,6 +22,30 @@ func HandleHealth() http.HandlerFunc {
...
+		Timeout: 1 * time.Second,
+	}
+	conn, err := dialer.Dial("tcp", "8.8.8.8:53")
+
+	if err != nil {
+		logrus.Infoln("Error connecting to 8.8.8.8:53", err)
+		return false
 	}
+	defer conn.Close()
```
Comment (raghavharness) [inline handler/health.go:49]: Add this on line 44.

File: handler/health.go (hunk 1 above)
Comment (devkimittal) [reply handler/health.go:32]: sure will add a new field message in  HealthResponse

File: handler/health.go (hunk 2 above)
Comment (devkimittal) [reply handler/health.go:40]: we can increase the timeout to 2 seconds.

File: cli/client/http.go [hunk 4]
```diff
@@ -193,6 +193,10 @@ func (c *HTTPClient) GetStepLogOutput(ctx context.Context, in *api.StreamOutputR
 
 func (c *HTTPClient) Health(ctx context.Context) (*api.HealthResponse, error) {
 	path := "healthz"
+	if IsHosted(ctx) {
+		path += "?hosted=true"
```
Comment (vistaarjuneja) [inline cli/client/http.go:197]: Use something more generic like `perform_dns_lookup`

File: cli/client/util.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+import "context"
+
+type ContextKey string
+
+const (
+	Hosted = ContextKey("hosted")
+)
+
+func IsHosted(ctx context.Context) bool {
```
Comment (vistaarjuneja) [inline cli/client/util.go:11]: Remove this file - it unnecessarily complicates things

File: cli/client/http.go [hunk 6]
```diff
@@ -193,6 +193,10 @@ func (c *HTTPClient) GetStepLogOutput(ctx context.Context, in *api.StreamOutputR
 
 func (c *HTTPClient) Health(ctx context.Context) (*api.HealthResponse, error) {
 	path := "healthz"
+	if IsHosted(ctx) {
```
[the remark this answers is not in the document]
Comment (devkimittal) [reply cli/client/http.go:196]: not adding anythign to context here. It was added in an earlier pr and we are just reusing that

File: cli/client/util.go (hunk 5 above)
Comment (devkimittal) [reply cli/client/util.go:11]: we need as a common file  between runner and lite engine

File: cli/client/http.go [hunk 7]
```diff
@@ -193,6 +193,10 @@ func (c *HTTPClient) GetStepLogOutput(ctx context.Context, in *api.StreamOutputR
 
 func (c *HTTPClient) Health(ctx context.Context) (*api.HealthResponse, error) {
```
[the remark this answers is not in the document]
Comment (devkimittal) [reply cli/client/http.go:194]: in any case we have to pick it up from context, why not pick here only (requires less change).

[the remark this answers is not in the document]
Comment (vistaarjuneja) [reply cli/client/http.go:194]: It should be easy to use this library for a client to interact with a server. Let's say a client wants to use `client/http.go` and wants to do a health check. From the definition, it's not clear what the options are and adding it to context is not intuitive.

File: cli/client/http.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vistaarjuneja) [reply cli/client/http.go:196]: You can move this to where you call the Health function.
```
  resp, err := client.Health(ctx, (IsHosted(ctx) && isWindows))
```

File: cli/client/http.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (devkimittal) [reply cli/client/http.go]: done

File: cli/client/util.go (hunk 5 above)
Comment (devkimittal) [reply cli/client/util.go]: removed

**Review by vistaarjuneja — APPROVED** [review]

**Pull request merged**